qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py

```diff
@@ -410,7 +419,19 @@ def simplify(self, atol=None, rtol=None):
             rtol = self.rtol
 
         # Filter non-zero coefficients
-        non_zero = np.logical_not(np.isclose(self.coeffs, 0, atol=atol, rtol=rtol))
+        if self.coeffs.dtype == object:
+            non_zero = np.logical_not(
+                [
+                    np.isclose(coeff, 0, atol=atol, rtol=rtol)
+                    if not hasattr(coeff, "sympify")
+                    else complex(coeff.sympify())
+                    if coeff.sympify().is_Number
+                    else False  # Symbol is not zero.
+                    for coeff in self.coeffs
+                ]
```

Comment:
This is pretty hard to read, and it causes objects that trigger `complex(coeff.sympify())` to get treated differently to other complex numbers.  You might want to refactor the inner part into a function, or do something like
```python
def to_complex(coeff):
    if not hasattr(coeff, "sympify"):
        return coeff
    sympified = coeff.sympify()
    return complex(sympified) if sympified.is_Number else np.nan

non_zero = np.logical_not(np.isclose([to_complex(x) for x in self.coeffs], 0, atol=atol, rtol=rtol))
```
to guarantee consistency.  Speed probably isn't a factor, since `object` dtype arrays lose all the vectorisation benefits anyway (I strongly imagine, since they'll need to acquire the GIL and run in Python space, but I haven't actually tested).

Reply:
Thank you for your suggestion. reflected in d01b52e.